### PR TITLE
chore: replace usages of AddressBook with Roster in Platform code

### LINKED ISSUE #14970: Replace AB with Roster: com.swirlds.platform

### PR DESCRIPTION
**Pull request by anthony-swirldslabs**
**Description**:
As a part of the TSS initiative, and specifically the TSS-Roster project, we're refactoring the Platform code to replace usages of `AddressBook` with `Roster`. This is a phase 1 of the refactoring that covers some 50% or maybe even 75% or more of the usages. Other places in code, such as the platform state and related entities, may require more complex changes and will be refactored separately in subsequent separate PRs.

Note that while technically it's outside of the platform code, some code in services/app had to be modified as well because it interacts with Platform classes that have been refactored as a part of the fix.

Please note that all the changes are interrelated as this whole diff has grown from modifying just a single class initially. It's impossible to remove some of the changes from this PR w/o introducing artificial Roster <-> AddressBook conversions in code which would affect the performance of the system. Therefore, unfortunately, it's impossible to break up this PR into smaller pieces.

**Related issue(s)**:

Fixes #14970

**Notes for reviewer**:
Tests should pass.

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/crypto/CryptoStatic.java [hunk 1]
```diff
@@ -232,6 +235,22 @@ public static X509Certificate generateCertificate(
...
     }
 
+    /**
+     * Generate a X509Certificate from a byte array that was previously obtained via X509Certificate.getEncoded().
+     *
+     * @param encoded a byte array with an encoded representation of a certificate
+     * @return the certificate reconstructed from its encoded form
+     */
+    @NonNull
+    public static X509Certificate generateCertificate(@NonNull final byte[] encoded) {
```
Comment (edward-swirldslabs) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/crypto/CryptoStatic.java:245]: Is this really generating a new certificate or decoding it from the encoding?    I think the term `generate` in cryptography means to make something new.

Comment (anthony-swirldslabs) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/crypto/CryptoStatic.java]: Good point. The method is now renamed to `decodeCertificate`.

**Review by edward-swirldslabs — COMMENTED** [review]
I started going glossy eyed when it got to the unit test.   It would be really helpful to create a list of files where the code was not just a 1-1 replacement to bring to attention where the code review is really needed.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/proof/algorithms/StateProofUtils.java [hunk 2]
```diff
@@ -43,43 +44,49 @@ private StateProofUtils() {}
...
+            if (address.weight() == 0) {
                 // Don't bother validating the signature of a zero weight node.
                 continue;
             }
 
-            if (!signatureVerifier.verifySignature(nodeSignature.signature(), hashBytes, address.getSigPublicKey())) {
+            if (!signatureVerifier.verifySignature(
+                    nodeSignature.signature(),
+                    hashBytes,
+                    CryptoStatic.generateCertificate(
```
[the remark this answers is not in the document]
Comment (anthony-swirldslabs) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/proof/algorithms/StateProofUtils.java]: It just got renamed in a commit I pushed an hour ago.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/connectivity/OutboundConnectionCreator.java [hunk 3]
```diff
@@ -131,23 +138,4 @@ public Connection createConnection(final NodeId otherId) {
...
     }
-
-    /**
-     * Find the best way to connect <code>from</code> address <code>to</code> address
-     *
-     * @param from the address that needs to connect
-     * @param to   the address to connect to
-     * @return the IP address to connect to
-     */
-    private String getConnectHostname(final Address from, final Address to) {
```
[the remark this answers is not in the document]
Comment (anthony-swirldslabs) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/connectivity/OutboundConnectionCreator.java:142]: With the removal of distinct "external" and "internal" endpoints and replacing them with a simple list, this has become a little complex. While I agree that there needs to be a more sophisticated solution to connect to a proper endpoint, I think that the complexity of the added logic may be a bit high for this refactoring chore. Do you want me to file a separate ticket to reimplement the endpoint choosing method in the new Roster world?

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/connectivity/OutboundConnectionCreator.java [hunk 4]
```diff
@@ -78,10 +78,17 @@ public OutboundConnectionCreator(
...
+        final RosterEntry other = RosterUtils.getRosterEntry(roster, otherId.id());
+        final RosterEntry ownRosterEntry = RosterUtils.getRosterEntry(roster, selfId.id());
+
+        // NOTE: we always connect to the first ServiceEndpoint, which for now represents a legacy "external" address
+        // (which may change in the future as new Rosters get installed).
+        // There's no longer a distinction between "internal" and "external" endpoints in Roster,
+        // and it would be complex and error-prone to build logic to guess which one is which.
+        // Ideally, this code should use a randomized and/or round-robin approach to choose an appropriate endpoint.
+        // For now, we default to the very first one at all times.
+        final int port = RosterUtils.fetchPort(other);
```
[the remark this answers is not in the document]
Comment (anthony-swirldslabs) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/connectivity/OutboundConnectionCreator.java:90]: Please see a comment on this topic below.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/recovery/EventRecoveryWorkflow.java [hunk 5]
```diff
@@ -160,10 +171,50 @@ public static void recoverState(
...
+                            + " keys, while there's " + rosterEntryMap.size() + " roster entries.");
+                }
+
+                allKeysAndCerts.forEach((nodeId, kc) -> {
+                    if (!rosterEntryMap.containsKey(nodeId.id())) {
+                        throw new IllegalStateException("There's not a roster entry for node id " + nodeId.id());
+                    }
+
+                    try {
+                        if (!Arrays.equals(
```
[the remark this answers is not in the document]
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/recovery/EventRecoveryWorkflow.java:194]: oh yeah! :-)

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/connectivity/OutboundConnectionCreator.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/connectivity/OutboundConnectionCreator.java:142]: Unfortunately, I think this has to be fixed in this PR. 

Here is a [mainnet.config.txt](https://github.com/swirlds/infrastructure/blob/cdcc7800538ec804858dbb96c0e4cfc3dc4ff46b/mainnet/ansible/config.mainnet.txt)

Multiple nodes list `10.x.x.x` IP addresses as internal addresses.   We will never successfully connect to these IP addresses if we attempt to connect to them.  We have to connect to the IP address in the external position.   If we don't fail over from attempting the first IP address, then we just fail and that isn't going to work.

**Review by edward-swirldslabs — CHANGES_REQUESTED** [review]
Need to block this PR until we figure out the Internal/External address solution.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/gossip/SyncGossip.java [hunk 6]
```diff
@@ -190,9 +191,16 @@ public SyncGossip(
...
         // allow other members to create connections to me
-        final Address address = addressBook.getAddress(selfId);
+        final RosterEntry rosterEntry = RosterUtils.getRosterEntry(roster, selfId.id());
+        // Assume all ServiceEndpoints use the same port and use the port from the first endpoint.
+        // Previously, this code used a "local port" corresponding to the internal endpoint,
+        // which should normally be the second entry in the endpoints list if it's obtained via
+        // a regular AddressBook -> Roster conversion.
+        // The assumption must be correct, otherwise, if ports were indeed different, then the old code
+        // using the AddressBook would never have listened on a port associated with the external endpoint,
+        // thus not allowing anyone to connect to the node from outside the local network, which we'd have noticed.
```
[the remark this answers is not in the document]
Comment (anthony-swirldslabs) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/gossip/SyncGossip.java:201]: This isn't a completely new behavior introduced by the refactoring though. The legacy code would just listen on the `Address.getListenPort()`, and the latter would return the result of `getPortInternal()`. In other words, the current code assumes that the port number is the same for both internal and external endpoints.

Since Roster doesn't differentiate between "internal" and "external" endpoints, this same assumptions simply transforms here to assuming that any endpoint defined in the RosterEntry uses the same port, and hence using the one from the entry with index 0 is equivalent to the old code using the internal port number.

So this refactoring doesn't really make this part any different from the current code behavior.

I do agree that this may not be the best strategy, especially if different endpoints may want to use different port numbers in the future, which would require a node to open several server sockets on different ports, and preferably bind them to specific IP addresses of the node per the service endpoint definitions. However, overhauling this part is completely outside the scope of this PR or the Roster or TSS projects as a whole. Please open a separate ticket if you feel strongly about the overhaul.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/connectivity/OutboundConnectionCreator.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (anthony-swirldslabs) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/connectivity/OutboundConnectionCreator.java:142]: Yes, I considered that.

All the `10.x.x.x` IP addresses are listed as the first addresses in config.txt, and the `AddressBookUtils.parseAddressText` correctly uses the first address as an internal address, and then uses the second address as an external one.

The `RosterRetriever.buildRoster` puts the external address into a ServiceEndpoint with index 0, and the internal address into the ServiceEndpoint with index 1.

The refactored code basically defaults to using the ServiceEndpoint with index 0 at all times - that is whenever it opens a server socket, or a client socket to connect to a peer.

What all this means is that the refactored code will only ever use the external IP of a node to connect to it. Unless a sub-network hosting several nodes at once has a firewall setup in such a way that it prohibits the members of the sub-net to connect to their external IP addresses, this should work at all times and allow all nodes to be able to connect to all the other nodes. I don't believe we have firewalls configured in such a way, but even if we do, resolving this would be a matter of reconfiguring the firewall to allow such connections.

Like a comment above in this same class states, it would be ideal to have a more sophisticated logic in order to allow the use of local ip addresses when possible. But given that the Roster doesn't explicitly differentiate between local and external addresses, building such a logic may be not trivial.

Long story short, per the above reasoning, the refactoring code should work. If it needs to be improved, this could probably be done in a separate ticket, perhaps even unrelated to the TSS/Roster specifically, since, as mentioned in the port number-related conversation above, it may need even a bigger overhaul.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/recovery/EventRecoveryWorkflow.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (anthony-swirldslabs) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/recovery/EventRecoveryWorkflow.java:194]: Thanks for confirming. I'm resolving this conversation then.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/connectivity/OutboundConnectionCreator.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/connectivity/OutboundConnectionCreator.java:142]: I am asking in the DevOps channel if they or Solo  is expecting to make connections through the `internal` ip address.   If not, then creating an issue to perform the correct behavior and implementing it later is fine.  But if this PR breaks behavior, then we need to implement it as part of this PR.

[the remark this answers is not in the document]
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/connectivity/OutboundConnectionCreator.java:142]: I've created an issue for this item and added it to the sprint.   https://github.com/hashgraph/hedera-services/issues/16065

**Review by litt3 — COMMENTED** [review]
> It's impossible to remove some of the changes from this PR w/o introducing artificial Roster <-> AddressBook conversions in code which would affect the performance of the system. Therefore, unfortunately, it's impossible to break up this PR into smaller pieces.

I don't really think this is true. This could be broken up by, for example, doing a single conversion from `AddressBook` to `Roster` in `PlatformComponentBuilder`, and then modifying the platform components one at a time to accept a roster instead of an address book. 

Without dynamic address book, a single conversion (or even a couple) when the node starts up isn't going to have any effect on performance.

**Comment by anthony-swirldslabs** [comment]
@litt3 :

> I don't really think this is true. This could be broken up by, for example, doing a single conversion from `AddressBook` to `Roster` in `PlatformComponentBuilder`, and then modifying the platform components one at a time to accept a roster instead of an address book.

I was unable to refactor the code this way because once a component starts using a Roster, when it communicates with other components, it would have to perform the conversions to AB on the fly. Also, I didn't want to pass both the Roster and the AB to every component until all of them migrate as this would leave more cleanup work for the future.

**Comment by litt3** [comment]
> I was unable to refactor the code this way because once a component starts using a Roster, when it communicates with other components, it would have to perform the conversions to AB on the fly. Also, I didn't want to pass both the Roster and the AB to every component until all of them migrate as this would leave more cleanup work for the future.

Which components are passing address books back and forth to each other, that would need on-the-fly conversion? All components that I can think of off the top of my head just accept an address book when they are constructed, and then use it internally for their own purposes. 

Take `TipsetEventCreator`, for instance. I think you can just pass in a `Roster` to its constructor, modify dependent tipset classes, and that would be an incremental step. Splitting this PR into 4 or 5 chunks like that would make it vastly more reviewable.

**Comment by anthony-swirldslabs** [comment]
> Take `TipsetEventCreator`, for instance. I think you can just pass in a `Roster` to its constructor, modify dependent tipset classes, and that would be an incremental step. Splitting this PR into 4 or 5 chunks like that would make it vastly more reviewable.

The `TipsetEventCreator` creates instances of `TipsetMetrics`, `TipsetTracker`, and `TipsetWeightCalculator`. The `TipsetTracker` then creates a `Tipset`, and so does the `TipsetWeightCalculator`. Then there's unit tests for all these classes. And there may (or may not) be other usages of these classes. All these would need to switch to using the Roster, and depending on what other classes they interact with, those other classes may need to do so as well.

I suppose this particular chunk of code could in theory be split into a separate PR, although it wouldn't be particularly small anyway given it would need to update at least 6 or 7 classes with all their dependencies. However, I haven't seen this chunk as a separate one because I didn't start this change with looking into the `TipsetEventCreator` specifically. As I mentioned in the description, this change originated elsewhere (in `Utilities.createPeerInfoList` IIRC), which subsequently got me to update all the code that is present in this PR.

**Comment by litt3** [comment]
> The `TipsetEventCreator` creates instances of `TipsetMetrics`, `TipsetTracker`, and `TipsetWeightCalculator`. The `TipsetTracker` then creates a `Tipset`, and so does the `TipsetWeightCalculator`. Then there's unit tests for all these classes. And there may (or may not) be other usages of these classes. All these would need to switch to using the Roster, and depending on what other classes they interact with, those other classes may need to do so as well.
> 
> I suppose this particular chunk of code could in theory be split into a separate PR, although it wouldn't be particularly small anyway given it would need to update at least 6 or 7 classes with all their dependencies. However, I haven't seen this chunk as a separate one because I didn't start this change with looking into the `TipsetEventCreator` specifically. As I mentioned in the description, this change originated elsewhere (in `Utilities.createPeerInfoList` IIRC), which subsequently got me to update all the code that is present in this PR.

My point is that it's not impossible to split this PR- there's a pretty easy way to split things up, which is by component.

**Comment by anthony-swirldslabs** [comment]
> My point is that it's not impossible to split this PR- there's a pretty easy way to split things up, which is by component.

You're correct that it's not impossible. However, it would be incorrect to assume that it's easy for someone unfamiliar with the code base. Could you please help and guide me with suggestions on what next "refactoring root class" I should pick up as I make incremental progress with smaller PRs? As the first step, I've just created the following PR based on your suggestion: 

https://github.com/hashgraph/hedera-services/pull/16102 

Please note that while it's substantially smaller than this one, I wouldn't call the new PR a small one still. But I'm unsure if splitting it further would be helpful or reasonable.

> I'm not asserting it must be split. Just that if I was a reviewer, I wouldn't feel comfortable with the proposed change set.

I totally agree with you regarding the PR size - I myself am the last person to tolerate or appreciate large PRs. But like I mentioned, my unfamiliarity with the codebase got me into this large change set. If you do agree to guide me as I'm doing this refactoring, hopefully we'll be able to release the refactoring in smaller and easier to digest chunks.

**Comment by litt3** [comment]
I think I might do groupings something like this:

- `DefaultEventSignatureValidator`
- `SyncGossip`
- `DefaultBranchReporter`
- `DefaultConsensusEngine`
- the demo applications

and then there would certainly be some misc changes, where address book is used here and there.

**Comment by anthony-swirldslabs** [comment]
Thank you, @litt3 ! I'll try and follow the list that you provided, and will reach out if I run into issues.

Since we're going to release this change in smaller pieces, I'm closing this huge PR.
